cpp/include/cudf/timezone.hpp

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <rmm/mr/device/per_device_resource.hpp>
+
+#include <memory>
+#include <optional>
+#include <string>
+
+namespace cudf {
+class table;
+
+// Cycle in which the time offsets repeat in Gregorian calendar
+static constexpr int32_t solar_cycle_years = 400;
+// Number of future entries in the timezone transition table:
+// Two entries per year, over the length of the Gregorian calendar's solar cycle
```

Comment:
How do we know there are two entries per year? Is it guaranteed by the spec that timezones cannot have zero, one, three, or four entries?

Reply:
There are two entries because of Daylight Saving Time. If the time zone does not have DST, all future entries will have the same time offset. We could optimize this case in the future.

Reply:
It seems that we’re hardcoding an assumption that daylight savings time exists and results in exactly two changes per year. I don’t think either of those are safe assumptions.

Reply:
It is a safe assumption, since that's how `make_timezone_transition_table` works (currently). It will always include two entires per year in cycle. The case when DST does not exist, it will create 2 X cycle_years of the same offset, which is still correct, just slower than necessary.
We could separate the output columns into tzif entires and future entries. With that, we can have any number of elements in the future column. I would like to keep optimizations like this to a future PR. Let me know what you think.

Reply:
Document this ("it always includes two entries per year, assuming two transitions per year due to Daylight Savings Time") and any potential future work in a comment. That would be enough to resolve this and move forward.

Reply:
Agree with the suggestion, this on not clear from the API.
Added more detailed description of the transition table.